feat: Streamline `sentry-trace`, `baggage` and DSC handling

mydea opened the pull request:
As a first step for https://github.com/getsentry/sentry-javascript/issues/12385, I looked though our current implementations, and tried to streamline this. We have a bunch of somewhat duplicate code there that handles sentry-trace & baggage headers _mostly_ the same but not _fully_ the same, as well as relatedly the DSC.

To streamline this, I added some new methods in core:

* `getTraceData` already exists, but was extended to also allow to pass a specific `span` to it. I updated some code to use this method that hasn't used it before, and also added some more tests - also around the ACS and the otel-specific implementation for this. 
* For this and edge cases, there are new primitives `getDynamicSamplingContextFromScope` and `getTraceContextFromScope` which handle picking these things from given scope etc.

While working on this, I also noticed that `captureCheckIn` in ServerRuntimeClient was actually not properly working in Node (OTEL), as it relied on the core way of scope-span coupling. So I also updated this to actually work as expected.

mydea commented on packages/core/src/baseclient.ts:
```diff
@@ -689,30 +688,18 @@ export abstract class BaseClient<O extends ClientOptions> implements Client<O> {
       };
 
-      const trace = evt.contexts && evt.contexts.trace;
-      if (!trace && propagationContext) {
```
These checks were not really necessary, I believe - first, `propagationContext` cannot be empty here. And second, we do not need to check for `trace` as this will be overwritten anyhow by `...evt.contexts` if it already exists.

mydea commented on packages/browser/src/tracing/request.ts:
```diff
@@ -419,21 +413,9 @@ export function xhrCallback(
-  };
-
-  const sentryTraceHeader =
-    span && hasTracingEnabled() ? spanToTraceHeader(span) : generateSentryTraceHeader(traceId, spanId, sampled);
```
Here, we have checked for `span && hasTracingEnabled()`, while in fetch we did not check for `hasTracingEnabled()`. I opted to use this logic (just check span) everywhere now...

Lms24 reviewed (commented):
maybe I'm missing something obvious but why do we need `getSentryHeaders`? Doesn't `getTraceData` already do this?

mydea commented on packages/core/test/lib/utils/traceData.test.ts:
```diff
@@ -1,156 +1,280 @@
+  return client;
+}
 
 describe('getTraceData', () => {
```
these tests relied a lot on mocks, and since the underlying implementation was changed a bunch of this broke. So I rewrote these tests to use "real" stuff as much as possible, but tried to keep the scenarios we want to cover intact.

mydea commented on packages/opentelemetry/src/propagator.ts:
```diff
@@ -91,9 +88,14 @@ export class SentryPropagator extends W3CBaggagePropagator {
     const url = activeSpan && getCurrentURL(activeSpan);
+    // We only want to check the URL against tracePropagationTargets for outgoing request spans
+    // In other cases, we always inject the trace data
+    const shouldCheckUrl =
```
Without this, injecting metadata e.g. in remix (but possibly other places!) would fail, because in this scenario the active span may be e.g. the http.server span, which has a `http.url` attribute that we would check and which may fail to check against `tracePropagationTargets`, disabling injection and thus not generating any meta tags.

Now, we only check against the http.url if the span is a http.client span.

mydea commented:
OK @Lms24 and @lforst I have updated this to now use `getTraceData` everywhere & consistently. 
This uncovered a bunch of edge cases and problems, which I hopefully all fixed now. Sadly bundle size _increased_ instead of decreased :( but IMHO it's worth it because at least behavior should be consistent now...!

mydea commented:
Nevermind, still not working - need another round 😬

mydea commented:
@lforst & @Lms24 OK, Now _for real_ I think it should work.

I had to change the custom `getTraceData` implementation in OTEL to not use the propagator directly, because this also ensured that we checked the active span's url for trace propagation targets and possibly skip the sentry headers 😬 So instead, we now call the `getInjectionData` method from the propagator directly in `getTraceData`, which should ensure we have the same implementation but without the checks.

mydea commented:
> I think this is a good refactor! A bit concerned that bundle size is increasing although it looks like we removed more code than we added 😅

yeah this is really weird 😬 not really sure where this is coming from... I think I'd still do this and hope that some follow up refactoring can safe a few bytes there. Sometimes, the ways of bundle size are not easy to understand 😬

mydea commented:
@Lms24 so, after digging some more into why this increases bundle size, I believe it is because of this code in `getTraceData`:

```js
  const isValidSentryTraceHeader = TRACEPARENT_REGEXP.test(sentryTrace);
  if (!isValidSentryTraceHeader) {
    logger.warn('Invalid sentry-trace data. Cannot generate trace data');
    return {};
  }

  const validBaggage = isValidBaggageString(baggage);
  if (!validBaggage) {
    logger.warn('Invalid baggage data. Not returning "baggage" value');
  }
```

which basically checks some regexes. We have not checked this before, so now we have to include these regexes in the browser bundles which increases bundle size 😬 

So IMHO either we accept this, or we only check these things in debug mode (?). I am not aware of any problems we had with this so far in browser land where we did not use to check this, so I am missing context for why we are checking this so defensively here.

lforst commented:
Honestly checking for validity of shapes is a bit weird. Our parsing/writing logic should just have sensible defaults and there is no need to be so defensive. Can we remove this checking logic?

Lms24 commented:
I think it's fine to remove the validation logic!

mydea commented:
I will merge this as is for now, and we can look into the validation in a follow up PR 👍
